### ISSUE TITLE
Automate PR checking with Danger

### ISSUE THREAD

**Issue by rnystrom** [Instagram/IGListKit#394]
http://danger.systems/ cc @orta

### Warnings

- No CHANGELOG.md entry for any change
- Non-final class in examples
- Big PR
- [Swiftlint](https://github.com/ashfurrow/danger-swiftlint) for examples
- Markdown verification using [prose](https://github.com/dbgrandi/danger-prose) (incl. link checking)
- ObjC classes without `IG_SUBCLASSING_RESTRICTED`

### Failures

- New CHANGELOG.md entry when `.h` file is changed
- Facebook copyright in headers
- Only allow editing `/docs`from admins
- WIP pull requests
- Changes to restricted file types (`.lock`, `.xcconfig`, etc)
- **Always fail**: "Merging should be done internally"

Any other ideas?

**Comment by Sherlouk** [comment]
Failure on changes to Pod files (`.lock`, `.xcconfig`)

Source https://github.com/Instagram/IGListKit/pull/368#issuecomment-269092317

**Comment by jessesquires** [comment]
Danger plugins:

- Proselint: https://github.com/dbgrandi/danger-prose
- Swiftlint: https://github.com/ashfurrow/danger-swiftlint

**Comment by Sherlouk** [comment]
Could probably also do with a changelog format validator.
`{summary} {Name} (#{PR Number})`

Should be easy to at least validate the correct PR and PR link

**Comment by orta** [comment]
There's [a semi-standard](http://keepachangelog.com/en/0.3.0/) for CHANGELOGs too - this plugin implements it - http://danger.systems/plugins/changelog.html

looks like https://github.com/ruby-grape/grape/blob/master/CHANGELOG.md

**Comment by Sherlouk** [comment]
As discussed in #399, we should **always** fail with a message similar to "Merging should be done internally" to prevent the merge button from ever being usable.

@rnystrom Should I go ahead and update the original comment with the various rules (as discussed above in thread) so we can tick them off once done?

**Comment by rnystrom** [comment]
@Sherlouk go for it!

Some linter ideas:

- Error on Swift classes in the examples marked `final`
- Warn on ObjC classes w/out using `IG_SUBCLASSING_RESTRICTED`

**Comment by Sherlouk** [comment]
@rnystrom Assume you meant classes not marked `final` for Swift?

**Comment by Sherlouk** [comment]
@rnystrom Worth keeping this issue open as there are still unimplemented rules?

**Comment by rnystrom** [comment]
@Sherlouk good call!

**Comment by jessesquires** [comment]
updated original issue to use markdown checklists 😄

**Comment by heshamsalman** [comment]
@jessesquires taking this on re: SwiftLint. 
I'm assuming y'all also want the script running in the example build phases? I have a `swiftlint.yml` that I pop into all of my projects -- should i start there, or with the default ruleset:  https://gist.github.com/heshamsalman/0910500186d4372f0c0d52ac73ea2110

**Comment by jessesquires** [comment]
thanks @heshamsalman !

let's start with the default rules 😄

**Comment by heshamsalman** [comment]
Heads up, there's gonna be a clean-up task immediately after this for the examples. Staring down the barrel of 228 warnings with default rules.

**Comment by jessesquires** [comment]
@heshamsalman - No worries. Let's do it in 1 PR with individual commits for each step so we can review it easier. 1 commit for swiftLint, 1 commit to fixup each example

**Comment by rnystrom** [comment]
@jessesquires @heshamsalman is there an "ObjC lint"? This is so cool.

**Comment by heshamsalman** [comment]
@rnystrom there are a few different objective c linters around -- I know of OCLint but haven't used it. 

I'll look into it and report back

**Comment by orta** [comment]
Not strictly a linter, but we use Space commander by square for the syntax parts of it

**Comment by Sherlouk** [comment]
Do we know if there's a way to have Travis and Danger as separate integrations on GitHub?

Would be nice to know if a build failed because of tests/etc vs Danger without having to go through the build log!

**Comment by orta** [comment]
Get Danger to print your test results?

**Comment by orta** [comment]
Or use Circle CI for Danger

**Comment by heshamsalman** [comment]
@rnystrom [Infer](http://fbinfer.com/) might be a good candidate here. Seems like someone (probably me) is gonna have to make a Danger plugin. 👍

**Comment by heshamsalman** [comment]
OCLint seems pretty similar to Swiftlint. Not quite as user friendly, but great overall.

**Comment by heshamsalman** [comment]
@rnystrom thoughts re: using xcov & danger-xcov to mandate a certain level of test coverage in production-facing files?

**Comment by Sherlouk** [comment]
So it turns out prose is a lot more annoying to do than I was first hoping...

Not only do you `danger-prose`, but you also need `proselint` and `mdspell`. To get `proselint` you also need `pip`.

Lots of dependencies which we know Travis doesn't really like (and would add quite a lot of time to builds until caching is fully implemented #757)

**Comment by heshamsalman** [comment]
Something tells me `pip` dependencies lead to a `six` error on travis.

**Comment by orta** [comment]
Yeah - there wasn't anything like `proselint` or `mdspell` in ruby, shame :)
